### PR TITLE
Improve padding behaviour of `ClapFeatureExtractor`

### LINKED ISSUE #23648: Unexpected padding behaviour of `ClapFeatureExtractor`
### System Info

- `transformers` version: 4.29.2
- Platform: Linux-4.15.0-204-generic-x86_64-with-glibc2.31
- Python version: 3.10.10
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 1.13.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Adding `padding=True` argument to `ClapFeatureExtractor` changes the padding strategy from `repeatpad`, which is the default, to constant padding ([code](https://github.com/huggingface/transformers/blob/04ab5605fbb4ef207b10bf2772d88c53fc242e83/src/transformers/models/clap/feature_extraction_clap.py#L248)).

Code adapted from: https://huggingface.co/docs/transformers/model_doc/clap#transformers.ClapModel.forward.example

```python
from datasets import load_dataset
from transformers import AutoProcessor

# load data
dataset = load_dataset("ashraq/esc50")
audio_sample = dataset["train"]["audio"][0]["array"]

# load data processor
processor = AutoProcessor.from_pretrained("laion/clap-htsat-unfused")

# pre-process data
inputs1 = processor.feature_extractor(audio_sample, return_tensors="pt")
inputs2 = processor.feature_extractor(audio_sample, return_tensors="pt", padding=True)

print((inputs1["input_features"] - inputs2["input_features"]).max())
# Output: tensor(119.4260)
```

This becomes a problem, for instance, when using `ClapProcessor`. `ClapProcessor` shares `kwargs` between the `tokenizer` and the `feature_extractor` ([code](https://github.com/huggingface/transformers/blob/04ab5605fbb4ef207b10bf2772d88c53fc242e83/src/transformers/models/clap/processing_clap.py#LL87C9-L87C9)). When using text inputs of different length, you need to pass `padding=True` argument to the `tokenizer`, but doing so changes the behaviour of the `feature_extractor`.

### Expected behavior

1. Either don't allow `padding=True` argument. Assert its value to be one of the allowed values - `repeatpad`, `repeat`, and `pad` in case of `ClapFeatureExtractor`.
2. Or, use the default padding strategy if `padding=True`.

As for sharing `kwargs`, I don't think that's a good idea. Would having two arguments, `tokenizer_kwargs` and `feature_extractor_kwargs` be better?

### PR DESCRIPTION
**Pull request by anmolojha**
# What does this PR do?
Currently, `ClapFeatureExtractor`'s padding behaviour is unintuitive. Specifically, 
* It works with any value of argument `padding` without any check. 
* It uses non default padding strategy if `padding=True`.

This was discussed in more detail around [this](https://github.com/huggingface/transformers/issues/23648#issuecomment-1558586027) comment of issue #23648.

This PR closes #23648. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi

### CODE REVIEW

**Comment by anmolojha** [comment]
Hi @sanchit-gandhi! I need your advice on how can we create this PR without breaking anything. I have implemented what seems like the best way to me. Looking forward to hearing from you. Thanks!

File: src/transformers/models/clap/feature_extraction_clap.py [hunk 1]
```diff
@@ -238,15 +238,23 @@ def _get_input_mel(self, waveform: np.array, max_length, truncation, padding) ->
 
         else:
             longer = False
-            # only use repeat as a new possible value for padding. you repeat the audio before applying the usual max_length padding
             if waveform.shape[0] < max_length:
-                if padding == "repeat":
+                if padding is True:
```
Comment (sanchit-gandhi) [inline src/transformers/models/clap/feature_extraction_clap.py:242]: How about we activate the 'default' behaviour of padding if `padding="max_length"`? We usually use `max_length` to indicate that we want to pad to the longest element in the batch, so we can still do this, but with the default padding strategy for CLAP

=> since we always truncate to `max_length`, then pad to `max_length`, we should allow this as an option IMO. It also prevents a breaking change

**Review by sanchit-gandhi — COMMENTED** [review]
Looks in good shape! Just a minor suggestion about setting a default value if a `max_length` padding strategy is used

File: src/transformers/models/clap/feature_extraction_clap.py (hunk 1 above)
Comment (anmolojha) [reply src/transformers/models/clap/feature_extraction_clap.py:242]: I think docstring makes it clear that the allowed values for padding are `repeatpad`, `repeat`, and `pad`.

https://github.com/huggingface/transformers/blob/5936c8c57ccb2bda3b3f28856a7ef992c5c9f451/src/transformers/models/clap/feature_extraction_clap.py#L284-L288

We can still allow `padding=True` and perform default padding in that case, but I am not sure of doing default padding for `padding="max_length"`.

Firstly, `max_length` does not mean the same thing here that it usually do. As you mentioned, `max_length`, generally means padding to the same length as longest sequence in a batch. But over here, all padding strategies map the input to the same length.
Secondly, `max_length` was never recommended to be used. Is it better to raise an error then doing something that the user wasn't expecting?

Let me know what you think.

**Comment by ArthurZucker** [comment]
Hey @anmolojha do you still want to work on this or should we close it? 🤗
